File: packages/module/jsonc-edit/src/parse.ts
import type { JsoncComment, } from './comment.ts';
import { JsoncParseError, } from './errors.ts';
import {
  closeArray,
  closeRecord,
} from './parse-close.ts';
import {
  parseScalar,
  type ValueScan,
} from './parse-scalar.ts';
import {
  appendComments,
  captureTrailing,
  prependComments,
  skipTrivia,
} from './parse-trivia.ts';
import { scanString, } from './scan.ts';
import type {
  JsoncKey,
  JsoncValue,
} from './value.ts';

//region Constants

/**
 * Maximum container nesting depth. Structural recursion is bounded so a
 * deeply-nested attacker-controlled document cannot overflow the call stack;
 * beyond this the parser throws rather than crashing.
 */
const MAX_DEPTH = 512;

//endregion Constants

//region Parser

/**
 * Parses one JSONC value starting at `index`, recursing into containers and
 * delegating scalars. Comments strictly inside the value are attached here; the
 * leading and trailing comments around it are attached by the caller.
 *
 * @param source - Full JSONC source.
 *
 * @param index - Offset of the value's first character.
 *
 * @param depth - Current nesting depth, guarded against overflow.
 *
 * @returns Parsed node and end offset.
 *
 * @throws JsoncParseError on malformed input or excessive nesting.
 *
 * @example
 * ```ts
 * parseValue({ source: 'true', index: 0, depth: 0 });
 * // => { node: { kind: 'boolean', value: true }, end: 4 }
 * ```
 */
export function parseValue({
  source,
  index,
  depth,
}: {
  readonly source: string;
  readonly index: number;
  readonly depth: number;
},): ValueScan {
  if (depth > MAX_DEPTH)
    throw new JsoncParseError({
      message: 'nesting too deep',
      offset: index,
    },);

  /**
   * First character of the value, selecting container versus scalar parsing.
   */
  const char = source[index];
  if (char === '{')
    return parseRecord({
      source,
      index,
      depth,
    },);
  if (char === '[')
    return parseArray({
      source,
      index,
      depth,
    },);
  return parseScalar({
    source,
    index,
  },);
}

/**
 * Parses an array body starting at the `[`. Elements carry their leading and
 * trailing comments; a trailing comma before the close is tolerated.
 *
 * @param source - Full JSONC source.
 *
 * @param index - Offset of the opening bracket.
 *
 * @param depth - Current nesting depth.
 *
 * @returns Parsed array node and end offset.
 *
 * @throws JsoncParseError when the array is malformed or unterminated.
 *
 * @example
 * ```ts
 * parseArray({ source: '[1, 2,]', index: 0, depth: 0 });
 * // => { node: { kind: 'array', elements: [...] }, end: 7 }
 * ```
 */
function parseArray({
  source,
  index,
  depth,
}: {
  readonly source: string;
  readonly index: number;
  readonly depth: number;
},): ValueScan {
  /**
   * Accumulated element nodes, built in place during the scan.
   */
  const elements: JsoncValue[] = [];
  for (let cursor = index + 1; ;) {
    /**
     * Leading comments and next significant offset.
     */
    const lead = skipTrivia({
      source,
      index: cursor,
    },);
    cursor = lead.end;
    if (cursor >= source.length)
      throw new JsoncParseError({
        message: 'unterminated array',
        offset: index,
      },);
    if (source[cursor] === ']')
      return {
        node: closeArray({
          elements,
          dangling: lead.comments,
        },),
        end: cursor + 1,
      };
    /**
     * Element value scanned at the cursor.
     */
    const valueScan = parseValue({
      source,
      index: cursor,
      depth: depth + 1,
    },);
    cursor = valueScan.end;
    /**
     * Trailing same-line comments and comma after the element.
     */
    const trailing = captureTrailing({
      source,
      index: cursor,
    },);
    cursor = trailing.end;
    elements.push(
      appendComments({
        node: prependComments({
          node: valueScan.node,
          comments: lead.comments,
        },),
        comments: trailing.comments,
      },),
    );
    if (!trailing.commaSeen) {
      /**
       * Lookahead confirming only a close may follow when no comma was seen.
       */
      const peek = skipTrivia({
        source,
        index: cursor,
      },);
      if (source[peek.end] !== ']')
        throw new JsoncParseError({
          message: 'expected , or ] in array',
          offset: peek.end,
        },);
    }
  }
}

/**
 * Parses an object body starting at the `{`. Each key carries the comment that
 * precedes it; each value carries the comment between colon and value plus its
 * trailing comment. Duplicate keys are preserved as separate entries, and a
 * trailing comma before the close is tolerated.
 *
 * @param source - Full JSONC source.
 *
 * @param index - Offset of the opening brace.
 *
 * @param depth - Current nesting depth.
 *
 * @returns Parsed record node and end offset.
 *
 * @throws JsoncParseError when the object is malformed or unterminated.
 *
 * @example
 * ```ts
 * parseRecord({ source: '{"a":1,}', index: 0, depth: 0 });
 * // => { node: { kind: 'record', entries: [...] }, end: 8 }
 * ```
 */
function parseRecord({
  source,
  index,
  depth,
}: {
  readonly source: string;
  readonly index: number;
  readonly depth: number;
},): ValueScan {
  /**
   * Accumulated key-to-value entries.
   */
  const entries: {
    key: JsoncKey;
    value: JsoncValue
  }[] = [];
  for (let cursor = index + 1; ;) {
    /**
     * Leading comments before the key, and next significant offset.
     */
    const lead = skipTrivia({
      source,
      index: cursor,
    },);
    cursor = lead.end;
    if (cursor >= source.length)
      throw new JsoncParseError({
        message: 'unterminated object',
        offset: index,
      },);
    if (source[cursor] === '}')
      return {
        node: closeRecord({
          entries,
          dangling: lead.comments,
        },),
        end: cursor + 1,
      };
    if (source[cursor] !== '"')
      throw new JsoncParseError({
        message: 'expected string key or } in object',
        offset: cursor,
      },);
    /**
     * Parsed entry, with cursor advanced past its value and trailing comma.
     */
    const entry = parseEntry({
      source,
      index: cursor,
      depth,
      leadComments: lead.comments,
    },);
    cursor = entry.end;
    entries.push({
      key: entry.key,
      value: entry.value,
    },);
    if (!entry.commaSeen) {
      /**
       * Lookahead confirming only a close may follow when no comma was seen.
       */
      const peek = skipTrivia({
        source,
        index: cursor,
      },);
      if (source[peek.end] !== '}')
        throw new JsoncParseError({
          message: 'expected , or } in object',
          offset: peek.end,
        },);
    }
  }
}

/**
 * Parses one `"key": value` entry, attaching the comment before the key, any
 * comment between key and colon, the comment between colon and value, and the
 * value's trailing comment.
 *
 * @param source - Full JSONC source.
 *
 * @param index - Offset of the key's opening quote.
 *
 * @param depth - Current nesting depth.
 *
 * @param leadComments - Comments that preceded the key.
 *
 * @returns Key, value, whether a comma followed, and the end offset.
 *
 * @throws JsoncParseError when the colon is missing.
 *
 * @example
 * ```ts
 * parseEntry({ source: '"a": 1', index: 0, depth: 0, leadComments: [] });
 * // => { key: {...}, value: {...}, commaSeen: false, end: 6 }
 * ```
 */
function parseEntry({
  source,
  index,
  depth,
  leadComments,
}: {
  readonly source: string;
  readonly index: number;
  readonly depth: number;
  readonly leadComments: readonly JsoncComment[];
},): {
  readonly key: JsoncKey;
  readonly value: JsoncValue;
  readonly commaSeen: boolean;
  readonly end: number;
} {
  /**
   * Scanned key string token.
   */
  const keyScan = scanString({
    source,
    index,
  },);
  /**
   * Comments sitting between the key and its colon.
   */
  const beforeColon = skipTrivia({
    source,
    index: keyScan.end,
  },);
  if (source[beforeColon.end] !== ':')
    throw new JsoncParseError({
      message: 'expected : after object key',
      offset: beforeColon.end,
    },);
  /**
   * Comments sitting between the colon and the value.
   */
  const beforeValue = skipTrivia({
    source,
    index: beforeColon.end + 1,
  },);
  /**
   * Value scanned after the colon.
   */
  const valueScan = parseValue({
    source,
    index: beforeValue.end,
    depth: depth + 1,
  },);
  /**
   * Trailing same-line comments and comma after the value.
   */
  const trailing = captureTrailing({
    source,
    index: valueScan.end,
  },);
  /**
   * Bare key node before its leading and inter-colon comments attach.
   */
  const baseKey: JsoncKey = {
    value: keyScan.value,
    raw: keyScan.raw,
  };
  return {
    key: appendComments({
      node: prependComments({
        node: baseKey,
        comments: leadComments,
      },),
      comments: beforeColon.comments,
    },),
    value: appendComments({
      node: prependComments({
        node: valueScan.node,
        comments: beforeValue.comments,
      },),
      comments: trailing.comments,
    },),
    commaSeen: trailing.commaSeen,
    end: trailing.end,
  };
}

//endregion Parser

File: package/rust-module/jsonc-edit/src/parse.rs
//! What: Iterative JSONC structural parsing with explicit container frames.
//! Why: Native Rust test threads overflowed their stacks before 512 nested containers returned.
//!
//! In TS you'd write (pseudocode):
//! ```ts
//! function parseJsonc(source: string): JsoncValue { const frames: Frame[] = []; /* loop over tokens */ }
//! ```

/// What: Import comment attachment for ordered parser trivia.
/// Why: Each key or value owns one normalized comment, not a flat token list.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// import { attach } from './comment';
/// ```
use crate::comment_merge::attach;
/// What: Import the source-borrowing scanner and same-line trivia result.
/// Why: Parser phases share one byte cursor.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// import { Scanner } from './scan';
/// ```
use crate::scan::Scanner;
/// What: Import JSON value, key, member, and parse-error types.
/// Why: Completed frame values move into their parent without decoding again.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// import type { JsoncEntry, JsoncKey, JsoncValue, JsoncParseError, JsoncKind } from './value';
/// ```
use crate::error::JsoncParseError;
/// What:     Import the document model types the parser builds.
/// Why:      Completed frames move into their parents, so the parser names the same types callers hold.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// import type { JsoncComment, JsoncEntry, JsoncKey, JsoncKind, JsoncValue } from './value';
/// ```
use crate::value::{JsoncComment, JsoncEntry, JsoncKey, JsoncKind, JsoncValue};

/// What: Maximum number of simultaneously open JSONC containers.
/// Why: Erroring before a 513th opener bounds the explicit parser frame list.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// const MAX_CONTAINERS = 512;
/// ```
const MAX_CONTAINERS: usize = 512;

/// What: The grammar phase of one open array or record frame.
/// Why: Separator enforcement must not be inferred from a comment or cursor alone.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// type Phase = 'item-or-close' | 'awaiting-child' | 'separator-or-close';
/// ```
#[derive(Clone, Copy, PartialEq, Eq)]
enum Phase {
    /// An item or close delimiter may follow.
    ItemOrClose,
    /// A nested child has started; the frame waits for its completed node.
    AwaitingChild,
    /// An item completed without a comma; a separator or close must follow.
    SeparatorOrClose,
}

/// What: One open array or record, including pending key and leading comments.
/// Why: A growable work stack replaces recursive calls while retaining source order.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// type Frame = { kind: 'array'; elements: JsoncValue[] } | { kind: 'record'; entries: JsoncEntry[]; key?: JsoncKey };
/// ```
enum Frame {
    /// Open array with completed elements and comments preceding its opener.
    ///
    /// What:     holds the elements parsed so far, the comments seen before the opener, and the
    ///           grammar phase.
    /// Why:      an array frame must know whether a value, a comma or the closer may come next.
    ///
    /// In TS you'd write (pseudocode):
    /// ```ts
    /// { kind: 'array', elements: JsoncValue[], leading: JsoncComment[], phase: Phase }
    /// ```
    Array {
        /// What:    Elements completed so far, in source order.
        /// Why:     `elements` stores finished children so the closer can build one value at the end.
        ///
        /// In TS you'd write (pseudocode):
        /// ```ts
        /// elements: JsoncValue[];
        /// ```
        elements: Vec<JsoncValue>,
        /// What:    Comments seen before this array's opening bracket.
        /// Why:     `leading` stores them until the array value exists, so they can be attached to it.
        ///
        /// In TS you'd write (pseudocode):
        /// ```ts
        /// leading: JsoncComment[];
        /// ```
        leading: Vec<JsoncComment>,
        /// What:    Grammar phase this frame expects next.
        /// Why:     `phase` stores the expectation, so a separator cannot be inferred from a comment
        ///          or cursor position alone.
        ///
        /// In TS you'd write (pseudocode):
        /// ```ts
        /// phase: Phase;
        /// ```
        phase: Phase,
    },
    /// Open record with ordered entries and at most one key awaiting its value.
    ///
    /// What:     holds the members parsed so far, the comments before the opener, a key awaiting its
    ///           value, and the grammar phase.
    /// Why:      a member's key and value carry separate comments, so the frame keeps them apart until
    ///           the value is complete.
    ///
    /// In TS you'd write (pseudocode):
    /// ```ts
    /// { kind: 'record', entries: JsoncEntry[], leading: JsoncComment[], key?: JsoncKey, phase: Phase }
    /// ```
    Record {
        /// What:    Members completed so far, in source order.
        /// Why:     `entries` stores finished pairs so duplicate keys stay visible rather than
        ///          collapsing into a map.
        ///
        /// In TS you'd write (pseudocode):
        /// ```ts
        /// entries: JsoncEntry[];
        /// ```
        entries: Vec<JsoncEntry>,
        /// What:    Comments seen before this record's opening brace.
        /// Why:     `leading` stores them until the record value exists, so they attach to the record.
        ///
        /// In TS you'd write (pseudocode):
        /// ```ts
        /// leading: JsoncComment[];
        /// ```
        leading: Vec<JsoncComment>,
        /// What:    The key whose value has not been parsed yet, if any.
        /// Why:     `pending_key` stores it so the value can be paired with its own key comment.
        ///
        /// In TS you'd write (pseudocode):
        /// ```ts
        /// key?: JsoncKey;
        /// ```
        pending_key: Option<JsoncKey>,
        /// What:    Grammar phase this frame expects next.
        /// Why:     `phase` stores the expectation, so a value cannot be accepted where a comma belongs.
        ///
        /// In TS you'd write (pseudocode):
        /// ```ts
        /// phase: Phase;
        /// ```
        phase: Phase,
    },
}

/// What:     Read and update one frame's grammar phase and closer.
/// Why:      Both frame variants carry a phase, so the driver asks the frame instead of matching on its
///           shape at every step.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// class Frame { get phase(): Phase; set phase(next: Phase); closeByte(): number }
/// ```
impl Frame {
    /// What: Read the frame's current grammar phase.
    /// Why: The driver can decide whether a value or a separator is legal.
    ///
    /// In TS you'd write (pseudocode):
    /// ```ts
    /// function phase(frame: Frame): Phase;
    /// ```
    fn phase(&self) -> Phase {
        // What: `match` extracts one of the two structural variants.
        // Why: Both variants store a phase but no caller should inspect their fields directly.
        //
        // In TS you'd write (pseudocode):
        // ```ts
        // return frame.phase;
        // ```
        return match self {
            Frame::Array { phase, .. } | Frame::Record { phase, .. } => *phase,
        };
    }

    /// What: Change the frame's expected next grammar event.
    /// Why: Accepting a child or comma must advance exactly one continuation.
    ///
    /// In TS you'd write (pseudocode):
    /// ```ts
    /// frame.phase = next;
    /// ```
    fn set_phase(&mut self, next: Phase) {
        match self {
            Frame::Array { phase, .. } | Frame::Record { phase, .. } => *phase = next,
        }
    }

    /// What: Select the matching close delimiter for this exact frame.
    /// Why: A `]` may not close an object or search for an older array.
    ///
    /// In TS you'd write (pseudocode):
    /// ```ts
    /// return frame.kind === 'array' ? ']'.charCodeAt(0) : '}'.charCodeAt(0);
    /// ```
    fn close_byte(&self) -> u8 {
        return if let Frame::Array { .. } = self { b']' } else { b'}' };
    }
}

/// What: Put newly encountered leading comments before comments already on a node.
/// Why: `/*outer*/[/*inner*/]` must retain source order even for empty containers.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// function prepend(node: JsoncValue, leading: JsoncComment[]): JsoncValue;
/// ```
fn prepend(node: &mut JsoncValue, leading: Vec<JsoncComment>) {
    // What: `take()` moves the existing optional comment out and leaves absence behind.
    // Why: JsoncComment merging can take ownership without duplicating a potentially long body.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // const inside = node.comment; node.comment = undefined;
    // ```
    let inside = node.comment.take();
    let prefixed = attach(None, leading);
    node.comment = if let Some(existing) = inside {
        attach(prefixed, vec![existing])
    } else {
        prefixed
    };
}

/// What: Attach trivia just before a comma or close to the last child value.
/// Why: Interstitial comments survive even when a comma begins a later line.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// function attachLast(frame: Frame, comments: JsoncComment[]): void;
/// ```
fn attach_last(frame: &mut Frame, comments: Vec<JsoncComment>) {
    if comments.is_empty() {
        return;
    }
    if let Frame::Array { elements, .. } = frame {
        let Some(last) = elements.last_mut() else { unreachable!("separator requires array element"); };
        last.comment = attach(last.comment.take(), comments);
        return;
    }
    if let Frame::Record { entries, .. } = frame {
        let Some(last) = entries.last_mut() else { unreachable!("separator requires record member"); };
        last.value.comment = attach(last.value.comment.take(), comments);
    }
}

/// What: Finish a matched container and move its completed values into one node.
/// Why: An emitted node is delivered exactly once to its immediate parent.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// function finish(frame: Frame, beforeClose: JsoncComment[]): JsoncValue;
/// ```
fn finish(frame: Frame, before_close: Vec<JsoncComment>) -> JsoncValue {
    // What: `match` consumes one frame variant and moves its owned children.
    // Why: Closure never re-parses or clones nested values.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // if (frame.kind === 'array') return { kind: 'array', elements: frame.elements };
    // ```
    let (mut node, leading) = match frame {
        Frame::Array { mut elements, leading, .. } => {
            let own = if let Some(last) = elements.last_mut() {
                last.comment = attach(last.comment.take(), before_close);
                None
            } else {
                attach(None, before_close)
            };
            (JsoncValue { kind: JsoncKind::Array { elements }, comment: own }, leading)
        }
        Frame::Record { mut entries, leading, pending_key, .. } => {
            if pending_key.is_some() {
                unreachable!("closing record with pending key");
            }
            let own = if let Some(last) = entries.last_mut() {
                last.value.comment = attach(last.value.comment.take(), before_close);
                None
            } else {
                attach(None, before_close)
            };
            (JsoncValue { kind: JsoncKind::Record { entries }, comment: own }, leading)
        }
    };
    prepend(&mut node, leading);
    return node;
}

/// What: Begin a value, either yielding a scalar or pushing one container frame.
/// Why: Containers never call this function recursively, so stack use stays constant.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// function beginValue(scanner: Scanner, leading: JsoncComment[], frames: Frame[]): JsoncValue | undefined;
/// ```
fn begin_value(
    scanner: &mut Scanner<'_>,
    leading: Vec<JsoncComment>,
    frames: &mut Vec<Frame>,
    delivered: &mut Option<JsoncValue>,
) -> Result<(), JsoncParseError> {
    if scanner.byte() == Some(b'{') || scanner.byte() == Some(b'[') {
        if frames.len() >= MAX_CONTAINERS {
            return Err(scanner.error("JSONC nesting too deep"));
        }
        if scanner.consume(b'{') {
            frames.push(Frame::Record { entries: Vec::new(), leading, pending_key: None, phase: Phase::ItemOrClose });
        } else {
            scanner.consume(b'[');
            frames.push(Frame::Array { elements: Vec::new(), leading, phase: Phase::ItemOrClose });
        }
        return Ok(());
    }
    let kind = if scanner.byte() == Some(b'"') {
        let (units, raw) = scanner.string()?;
        JsoncKind::Text { units, raw }
    } else {
        // What: An absent byte gets an invalid-number sentinel, not a false numeric match.
        // Why: EOF should produce the ordinary expected-value error.
        //
        // In TS you'd write (pseudocode):
        // ```ts
        // const first = sourceBytes[offset] ?? 0;
        // ```
        let first = scanner.byte().unwrap_or(0);
        if first == b'-' || first.is_ascii_digit() {
            let (raw, identity) = scanner.number()?;
            JsoncKind::Number { raw, identity }
        } else if scanner.source[scanner.offset..].starts_with("true") {
            scanner.offset += 4;
            JsoncKind::Boolean { value: true }
        } else if scanner.source[scanner.offset..].starts_with("false") {
            scanner.offset += 5;
            JsoncKind::Boolean { value: false }
        } else if scanner.source[scanner.offset..].starts_with("null") {
            scanner.offset += 4;
            JsoncKind::Null
        } else {
            return Err(scanner.error("expected JSONC value"));
        }
    };
    // What: `Some` hands the completed scalar to the parent in the next loop pass.
    // Why: A scalar needs no container frame or recursive call.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // delivered = { kind, comment: merge(leading) };
    // ```
    *delivered = Some(JsoncValue { kind, comment: attach(None, leading) });
    return Ok(());
}

/// What: Deliver one completed child and consume only its same-line trailing trivia.
/// Why: A comma after a later newline belongs to the parent's separator phase.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// function receiveChild(parent: Frame, child: JsoncValue, scanner: Scanner): void;
/// ```
fn receive_child(scanner: &mut Scanner<'_>, frames: &mut [Frame], mut child: JsoncValue) -> Result<(), JsoncParseError> {
    let trailing = scanner.capture_trailing()?;
    child.comment = attach(child.comment, trailing.comments);
    let Some(parent) = frames.last_mut() else { unreachable!("child without parent"); };
    if parent.phase() != Phase::AwaitingChild {
        return Err(scanner.error("unexpected completed JSONC value"));
    }
    if let Frame::Array { elements, .. } = parent {
        elements.push(child);
    } else if let Frame::Record { entries, pending_key, .. } = parent {
        let Some(key) = pending_key.take() else { unreachable!("record value without key"); };
        entries.push(JsoncEntry { key, value: child });
    }
    parent.set_phase(if trailing.comma { Phase::ItemOrClose } else { Phase::SeparatorOrClose });
    return Ok(());
}

/// What: Parse one complete JSONC document by driving a stack of container phases.
/// Why: Valid 512-depth input must not exhaust the Rust call stack.
///
/// In TS you'd write (pseudocode):
/// ```ts
/// function parseJsonc(source: string): JsoncValue { while (frames.length) advance(); }
/// ```
///
/// # Errors
/// Malformed syntax or a 513th open container returns a byte-positioned parse error.
pub fn parse_jsonc(source: &str) -> Result<JsoncValue, JsoncParseError> {
    // What: This mutable scanner borrows source and advances only through its own methods.
    // Why: The resulting tree owns text it needs after parsing ends.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // const scanner = { source, offset: 0 };
    // ```
    let mut scanner = Scanner::new(source);
    // What: `?` returns a scanner error early; a successful result owns ordered comments.
    // Why: Malformed leading trivia cannot silently disappear.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // const leading = skipTrivia(scanner);
    // ```
    let leading = scanner.trivia()?;
    if scanner.byte() != Some(b'{') && scanner.byte() != Some(b'[') {
        return Err(scanner.error("JSONC document root must be an object or array"));
    }
    let mut frames: Vec<Frame> = Vec::new();
    let mut delivered: Option<JsoncValue> = None;
    begin_value(&mut scanner, leading, &mut frames, &mut delivered)?;
    // What: The loop breaks with one completed root; every pass consumes input or a pending event.
    // Why: No recursion or retry of the same byte/phase pair is needed.
    //
    // In TS you'd write (pseudocode):
    // ```ts
    // let root: JsoncValue; while (frames.length || delivered) { /* advance */ }
    // ```
    let mut root = loop {
        if let Some(child) = delivered.take() {
            if frames.is_empty() {
                break child;
            }
            receive_child(&mut scanner, &mut frames, child)?;
            continue;
        }
        let Some(phase) = frames.last().map(Frame::phase) else {
            return Err(scanner.error("JSONC root did not complete"));
        };
        if phase == Phase::AwaitingChild {
            return Err(scanner.error("JSONC child did not complete"));
        }
        let before = scanner.trivia()?;
        if phase == Phase::SeparatorOrClose {
            let Some(top) = frames.last_mut() else { unreachable!("separator without frame"); };
            attach_last(top, before);
            if scanner.consume(b',') {
                top.set_phase(Phase::ItemOrClose);
                continue;
            }
            if scanner.byte() == Some(top.close_byte()) {
                let closer = top.close_byte();
                scanner.consume(closer);
                let Some(done) = frames.pop() else { unreachable!("close without frame"); };
                delivered = Some(finish(done, Vec::new()));
                continue;
            }
            return Err(scanner.error("expected comma or container close"));
        }
        let Some(top) = frames.last() else { unreachable!("item without frame"); };
        if scanner.byte() == Some(top.close_byte()) {
            let closer = top.close_byte();
            scanner.consume(closer);
            let Some(done) = frames.pop() else { unreachable!("close without frame"); };
            delivered = Some(finish(done, before));
            continue;
        }
        if let Some(Frame::Array { .. }) = frames.last() {
            let Some(top) = frames.last_mut() else { unreachable!("array frame disappeared"); };
            top.set_phase(Phase::AwaitingChild);
            begin_value(&mut scanner, before, &mut frames, &mut delivered)?;
            continue;
        }
        if scanner.byte() != Some(b'"') {
            return Err(scanner.error("expected double-quoted object key"));
        }
        let (units, raw) = scanner.string()?;
        let before_colon = scanner.trivia()?;
        if !scanner.consume(b':') {
            return Err(scanner.error("expected colon after JSONC key"));
        }
        let before_value = scanner.trivia()?;
        let key = JsoncKey { units, raw, comment: attach(attach(None, before), before_colon) };
        let Some(Frame::Record { pending_key, phase, .. }) = frames.last_mut() else {
            unreachable!("record frame disappeared");
        };
        *pending_key = Some(key);
        *phase = Phase::AwaitingChild;
        begin_value(&mut scanner, before_value, &mut frames, &mut delivered)?;
    };
    let trailing = scanner.trivia()?;
    if scanner.offset != source.len() {
        return Err(scanner.error("unexpected content after JSONC root"));
    }
    root.comment = attach(root.comment, trailing);
    return Ok(root);
}
